cmd/dockerd/daemon.go

```diff
@@ -228,7 +232,26 @@ func (cli *DaemonCli) start(opts *daemonOptions) (err error) {
 	// Notify that the API is active, but before daemon is set up.
 	preNotifyReady()
 
-	sp := setupTracing(ctx)
+	// Setup buildkit trace recorder
+	detect.Recorder = detect.NewTraceRecorder()
```

Comment:
Silly question; is there a specific reason to put this code in this part of the start-up cycle? Would there be benefits in setting things up as early as possible to trace as much of the flow as possible?

Reply:
No I don't think it would change anything.
This is for tracing builds which wouldn't have anything running at this point.

Reply:
Actually, I should likely move this below our tracing config... I'll have to double check what's happening in this function.

Reply:
Where it is is good.
All its doing:

```go
func NewTraceRecorder() *TraceRecorder {
	tr := &TraceRecorder{
		m:         map[trace.TraceID]*stubs{},
		listeners: map[trace.TraceID]int{},
	}

	go func() {
		t := time.NewTimer(60 * time.Second)
		for {
			<-t.C
			tr.gc()
			t.Reset(50 * time.Second)
		}
	}()

	return tr
}
```